### PR TITLE
Gradient Accumulation in Axlearn

### PR DESCRIPTION
**Pull request by apoorvtintin**
Gradient accumulation allows training with higher batch sizes without scaling out.

Added a new learner type ```learner.klass: 'axlearn.common.learner.AccumulatedLearner'```

At a high level the optimization does the following:
1. Input batch is split into even microbatches.
2. Creates a buffer for gradients and metrics.
3. Runs forward and backward pass for each microbatch in a loop summing up the gradients and aggregating metrics.
4. Average gradients across microbatches and normalize metrics.

Configuration changes:
- Number of microbatches are specified during configuration through option ```micriobatches``` in the learner.

### CODE REVIEW

**Review by ruomingp — COMMENTED** [review]
Could you explain why this is needed? We find usually it's more efficient to use either a larger mesh or a smaller batch size.

File: axlearn/common/learner.py [hunk 1]
```diff
@@ -94,8 +95,7 @@ def _prune_empty(in_tree: NestedTensor) -> NestedTensor:
     return prune_tree(in_tree, lambda _, v: isinstance(v, dict) and not v)
 
 
-@dataclasses.dataclass
-class ForwardOutputs:
+class ForwardOutputs(NamedTuple):
```
Comment (apghml) [inline axlearn/common/learner.py:98]: What is the purpose of this change?

File: axlearn/common/learner.py [hunk 2]
```diff
@@ -444,6 +444,153 @@ def _mask_tree(tree: dict, *, keep: dict) -> dict:
     )
 
 
+class MetricsAccumulationOp(NamedTuple):
```
Comment (apghml) [inline axlearn/common/learner.py:447]: Axlearn already has metric accumulation classes that are used by evalers. Could those be reused here instead of defining new classes?

File: axlearn/common/learner.py [hunk 3]
```diff
@@ -9,7 +9,8 @@
 
 import dataclasses
 import enum
-from typing import Callable, Mapping, Optional, Protocol, Sequence, Tuple
+import logging
```
Comment (apghml) [inline axlearn/common/learner.py:12]: We typically use absl logging instead of python logging.

File: axlearn/common/learner.py [hunk 4]
```diff
@@ -444,6 +444,153 @@ def _mask_tree(tree: dict, *, keep: dict) -> dict:
...
+    @config_class
+    class Config(Learner.Config):
+        """Configures Learner."""
+
+        # The number of microbatches to accumulate per optimizer step.
+        microbatches: Required[int] = REQUIRED
+        # tuple of key-value pairs specifying custom aggregation and normalization
+        # for a specific metric
+        metrics_accumulation_key_ops: Sequence[Dict[str, Optional[MetricsAccumulationOp]]] = []
+        gradient_dtype: Optional[jnp.dtype] = jnp.bfloat16
```
Comment (apghml) [inline axlearn/common/learner.py:493]: Does the existing learner class use this? If not, we should try to be consistent with its API.

File: axlearn/common/learner.py [hunk 5]
```diff
@@ -444,6 +444,153 @@ def _mask_tree(tree: dict, *, keep: dict) -> dict:
...
+        and metrics into buffers.
+
+        Args:
+            fn (ForwardFn): Model forward pass
+            inputs (NestedTensor): input microbatches with a leading microbatch axis
+            opt_params (NestedOptParam): params which don't need gradients
+
+        Returns:
+            ForwardBackwardOutputs: pytree containing gradients and metrics
+        """
```
Comment (apghml) [inline axlearn/common/learner.py:508]: I wonder if instead of having a separate learner for microbatching, it would be more flexible to have a generic way of wrapping a ForwardFn so that it uses Jax.lax.map to run the microbatches. Beyond avoiding the need to add a new learner, it would also allow for other microbetching uses outside of learner, eg inference or in second order optimizers.

**Comment by apghml** [comment]
I’m ooo this week. I have left some preliminary comments and will revisit on Monday.

File: axlearn/common/learner.py (hunk 1 above)
Comment (apoorvtintin) [reply axlearn/common/learner.py:98]: The interface of jax.lax.scan does not consider "_dataclass_" a valid JAX type.

File: axlearn/common/learner.py (hunk 5 above)
Comment (apoorvtintin) [reply axlearn/common/learner.py:508]: jax.lax.map gives no guarantees of sequential execution of microbatches which is the key quality of gradient accumulation.

File: axlearn/common/learner.py (hunk 2 above)
Comment (apoorvtintin) [reply axlearn/common/learner.py:447]: These metric accumulation classes are stateless to be usable as carry by jax.lax.scan unlike the ones in the evaler, I can make the class structure similar though

File: axlearn/common/learner.py (hunk 4 above)
Comment (apoorvtintin) [reply axlearn/common/learner.py:493]: Could you please be a little more specific, is the concern naming of variables?

**Comment by apghml** [comment]
Closing since gradient accumulation functionality has been implemented via https://github.com/apple/axlearn/pull/614
